### ISSUE TITLE
Saved scroll distance breaks when zoomed in

### ISSUE THREAD

**Issue by jpear1** [sumatrapdfreader/sumatrapdf#1438]
When SumatraPDF is closed and the current document is zoomed in, the saved scroll distance breaks. Usually when you reopen a document it will be scrolled to where it was when it was last opened, but I think it scales incorrectly with zoom because when I reopen a zoomed in document it is always scrolled to the last page.

**Comment by GitHubRulesOK** [comment]
Please confirm your version it worked in 3.1.2  and as far as I can tell is still working in 3.2.12085
It may help to supply a sample document if it only happens with some files
here I closed this file at this position and zoom
On reopening it is still the same, even if I close the session and start again
![image](https://user-images.githubusercontent.com/12754913/73308020-3f2fce00-4217-11ea-988c-92b3c5c195c5.png)

**Comment by jpear1** [comment]
Sorry, still new to this whole open source thing. Should have include all of this in my initial report. It has been updated.

**Comment by GitHubRulesOK** [comment]
@kjk its not time related as far as I can tell it fails instantly on reopening the file
I consider this as part of the "iceberg" issue of fitting files with varying widths thus although this exhibits as a different misbehaviour it is related to #1422

**Comment by kjk** [comment]
It happens in 3.1.2 so it's not related to fit-width changes.

A corner case so low priority.

**Comment by kjk** [comment]
Re-tested on current master and the saved position is restored exactly, so I'm closing this.

What I ran: a 20-page PDF where every 4th page is a different size (the varying-page-size case from #1422), `RememberStatePerDocument = true`, continuous mode. Zoomed to 200%, scrolled well into the document, closed SumatraPDF through File/Exit so the state was saved, reopened the same file: the vertical scroll position came back identical (difference of 0 pixels).

Note that a related bug in this area *was* real and is now fixed: in facing / book view, restoring a view whose first visible page was the second page of a row scrolled all the way right (#3591).

If you still see the position jump on the current [pre-release](https://www.sumatrapdfreader.org/prerelease), please reopen and attach the document — it may depend on the specific page sizes.

**Comment by jrpear** [comment]
Wonderful! I still use SumatraPDF, thank you maintainers for this excellent software :)

**Comment by jrpear** [comment]
Er, I just tried the pre-release version and it looks like it's still unfortunately broken on the `Album.of.Sonatinas.pdf` document I included in the original issue. The first time I scrolled somewhere, I don't remember, and the location was restored fine. But then after re-opening, I scrolled to page 9 at about 200% (so the page fit in the screen) and closed the document via the exit menu item and on re-open it took me to the top of the page! I scrolled down to the same place, closed SumatraPDF, re-launched it, and then SumatraPDF got stuck infinitely on "Loading Album of Sonatinas.pdf"...

@kjk

**Comment by kjk** [comment]
You're right, thanks for checking — I reproduced it with your `Album of Sonatinas.pdf` and it's now fixed in master.

**What happens:** at 200%, scrolled into page 9, File/Exit, reopen — the toolbar says page 9 but the canvas shows page 1. Worse, the position saved on *that* exit is `PageNo = 1`, so from then on the document keeps opening at the top. That matches what you described.

**Why:** your document has one page much wider than the others (the "iceberg" shape of #1422). The canvas is then as wide as that page and the normal pages sit centered in it, so scrolled fully left the viewport doesn't overlap *any* page. The code that answers "which page am I on" had a fallback for "no page is visible": page 1 if you're above page 1's bottom, otherwise the **last** page. Restoring a saved position runs straight through it:

```
SetScrollState(page 9, 134,29)
  GoToPage(9)                   -> viewPort.y = 26540   (correct)
  WM_SIZE -> SetViewPortSize    -> "current page" = 41, which it then restores
                                -> viewPort.y = 94194   (end of the document)
  CvtToScreen(page 9, ...)      -> y = -67574, computed against that wrong viewport
  GoToPage(9, -67574)           -> clamped to 0 = top of the document
```

The fallback now picks the page whose vertical band the viewport is in, which is what "current page" means in continuous mode. With that, your sequence round-trips exactly: `PageNo = 9, ScrollPos = 134.813 29.0629` is saved and comes back unchanged.

There's also a regression test now (`tests/issue-1438.ts`): a 12-page PDF with one 10x wider page, driven through two real sessions with their own settings directory. It fails on the old build and passes on the new one.

The fix will be in the next [pre-release](https://www.sumatrapdfreader.org/prerelease) build — please give it another go.

One thing I could not reproduce: the "stuck infinitely on Loading Album of Sonatinas.pdf" you saw afterwards. I opened the file a couple of dozen times without a hang. If it happens again, please open a separate issue and say whether the window is blank or shows the document, ideally with a screenshot.

**Comment by jrpear** [comment]
Yep, this no longer reproduces. The hang happened again with the update but after I cleared the history and opened it again the problem was gone. :)
